### PR TITLE
New AMC

### PR DESCRIPTION
**Pull request by nzmora**
Temp PR for Guy to review specifically:

examples/classifier_compression/utils_cnn_classifier.py
examples/classifier_compression/parser.py
examples/classifier_compression/utils_cnn_classifier.py
examples/classifier_compression/parallel-finetune.py

A more documented PR will follow after the review.

### CODE REVIEW

File: examples/classifier_compression/compress_classifier.py [hunk 1]
```diff
@@ -51,207 +51,93 @@
...
         #zeros_mask_dict = distiller.create_model_masks_dict(model)
         assert args.resumed_checkpoint_path is not None, \
             "You must use --resume-from to provide a checkpoint file to thinnify"
         distiller.remove_filters(model, compression_scheduler.zeros_mask_dict, args.arch, args.dataset, optimizer=None)
         apputils.save_checkpoint(0, args.arch, model, optimizer=None, scheduler=compression_scheduler,
                                  name="{}_thinned".format(args.resumed_checkpoint_path.replace(".pth.tar", "")),
                                  dir=msglogger.logdir)
-        print("Note: your model may have collapsed to random inference, so you may want to fine-tune")
-        return
+        print("Note: if your model collapsed to random inference, you may want to fine-tune")
```
Comment (guyjacob) [inline examples/classifier_compression/compress_classifier.py]: Use the instead of print?

File: examples/classifier_compression/compress_classifier.py [hunk 2]
```diff
@@ -51,207 +51,93 @@
...
+        do_exit = True
+    return do_exit
 
+def init_knowledge_distillation(args):
```
Comment (guyjacob) [inline examples/classifier_compression/compress_classifier.py]: Regardless of where this function ends up, the references to `model` in line 147 and `compression_scheduler` in line 148 are unresolved. Need to pass them to the function.

File: distiller/apputils/image_classifier.py [hunk 3]
```diff
@@ -0,0 +1,836 @@
...
+                                 name='_'.join([args.name, checkpoint_name]) if args.name else checkpoint_name,
+                                 dir=msglogger.logdir, extras={'quantized_top1': top1})
+
+def acts_quant_stats_collection(model, criterion, loggers, args):
+    msglogger.info('Collecting quantization calibration stats based on {:.1%} of test dataset'
+                   .format(args.qe_calibration))
+    model = distiller.utils.make_non_parallel_copy(model)
+    args.effective_test_size = args.qe_calibration
+    train_loader, val_loader, test_loader, _ = load_data(args)
+    test_fn = partial(test, test_loader=test_loader, criterion=criterion,
```
[the remark this answers is not in the document]
Comment (nzmora) [reply distiller/apputils/image_classifier.py]: done

File: examples/classifier_compression/compress_classifier.py [hunk 4]
```diff
@@ -51,207 +51,93 @@
...
 import distiller.apputils as apputils
-import distiller.model_summaries as model_summaries
-from distiller.data_loggers import *
 import distiller.quantization as quantization
-import examples.automated_deep_compression as adc
-from distiller.models import ALL_MODEL_NAMES, create_model
+from distiller.models import create_model
+#from utils_cnn_classifier import *
+import distiller.apputils.image_classifier as classifier
+from distiller.apputils.image_classifier import *
```
[the remark this answers is not in the document]
Comment (nzmora) [reply examples/classifier_compression/compress_classifier.py]: Done

File: examples/classifier_compression/parallel-finetune.py [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+    print("Fine-tuning checkpoint %s" % name)
+ 
+    app_args.gpus = str(gpu)
+    app_args.name = name
+    app_args.deprecated_resume = ckpt_file
+    app = classifier.ClassifierCompressor(app_args, script_dir=os.path.dirname(__file__))
+    app.train_loader, app.val_loader, app.test_loader = loaders
+    best = [float("-inf"), float("-inf"), float("inf")]
+    for epoch in range(app_args.epochs):
+        validate = epoch >= math.floor(VALIDATE_ENABLE_FACTOR * app_args.epochs)
```
[the remark this answers is not in the document]
Comment (nzmora) [reply examples/classifier_compression/parallel-finetune.py]: File `parallel-finetune.py` is currently tightly-coupled to AMC.  In this specific use-case, we are fine-tuning several checkpoints in parallel.  We are following a prescribed set of hyper-parameters and usually we don't care to follow this training process (i.e. we don't care to look at the loss, validation, training accuracies graphs).  So in essence we don't need to validate: we train a prescribed set of epochs and then take the test score.  But sometimes we want to set argument `--vs=0` and report the best validation score.  This is useful because sometimes the test score (==validation score in ImageNet when '--vs=0') of the last training epoch is not as good as some of the earlier epoch checkpoints.

I'm adding this code:
```
group.add_argument('--validate-enable-factor', type=float, default=0.2,
                           help="What portion of the epochs to validate (0=never validate; 1=validate after every "
                                "epoch; 0<factor<1 validate the last factor*ags.epochs epcohs."
                                "In the latter case, the reported score for the fine-tuned model is the "
                                "best performing validation score.")
:
:
    for epoch in range(app_args.epochs):
        validate = epoch >= math.floor((1-app_args.validate_enable_factor) * app_args.epochs)
        top1, top5, loss = app.train_validate_with_scheduling(epoch, validate=validate, verbose=False)
        if validate:
            if top1 > best[0]:
                best = [top1, top5, loss]
    if app_args.validate_enable_factor == 0:
        # We did not validate, so our score is the performance on the Test dataset
        return (name, app.test())
    return (name, best)
```

File: examples/classifier_compression/compress_classifier.py [hunk 6]
```diff
@@ -269,386 +155,24 @@ def main():
...
+    msglogger.info('=> using early-exit threshold values of %s', args.earlyexit_thresholds)
 
-    if args.quantize_eval:
-        model.cpu()
-        quantizer = quantization.PostTrainLinearQuantizer.from_args(model, args)
-        quantizer.prepare_model(distiller.get_dummy_input(input_shape=model.input_shape))
-        model.to(args.device)
 
-    top1, _, _ = test(test_loader, model, criterion, loggers, activations_collectors, args=args)
+class ClassifierCompressorSampleApp(classifier.ClassifierCompressor):
```
[the remark this answers is not in the document]
Comment (nzmora) [reply examples/classifier_compression/compress_classifier.py]: My plan is to subclass this (at a later time) so that, for example, we can have 
```
class EarlyExitSampleApp(ClassifierCompressorSampleApp):
:
```
which contains all of the EE-specific code.

File: examples/classifier_compression/compress_classifier.py [hunk 7]
```diff
@@ -51,207 +51,93 @@
...
-            optimizer = None
-            msglogger.info('\nreset_optimizer flag set: Overriding resumed optimizer and resetting epoch count to 0')
-
-    # Define loss function (criterion)
-    criterion = nn.CrossEntropyLoss().to(args.device)
+    args = parser.add_cmdline_args(classifier.init_classifier_compression_arg_parser()).parse_args()
+    app = ClassifierCompressorSampleApp(args, script_dir=os.path.dirname(__file__))
+    if app.handle_subapps():
+        return
+    init_knowledge_distillation(app.args)
```
[the remark this answers is not in the document]
Comment (nzmora) [reply examples/classifier_compression/compress_classifier.py]: Good point.  KD is currently half in the base-class and half in the derived class.  Ouch.
However, I want to move it _out_ of the base class.  That will require some refactoring that I don't want to do now, so I'm leaving things as is.  Users of the base class cannot use `--kd-teacher` anyhow.

File: examples/classifier_compression/parallel-finetune.py [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+        shutil.rmtree(app.logdir)
+        processes.append(FinetuningProcess(tasks, results, data_loader))
+        # Start the process
+        processes[-1].start()
+    
+    n_gpus = torch.cuda.device_count()
+
+    # Enqueue all of the fine-tuning tasks
+    for (instance, ckpt_file) in enumerate(checkpoints):
+        tasks.put(FinetuningTask(args=(ckpt_file, instance%n_gpus, app_args)))
```
[the remark this answers is not in the document]
Comment (nzmora) [reply examples/classifier_compression/parallel-finetune.py]: the current configuration covers the use-case you present.
The user can use as many processes as she wants.
When each process is _constructed_ it is assigned a GPU.  Once a process is constructed it is assigned multiple FT tasks, which all use the same GPU.
If you have 4 GPUs and you use 4 processes, each process uses a different GPU.  If you use 16 processes, GPU will be assigned to 4 different process.  E.g. I am able to train 16 CIFAR-10 networks in parallel (e.g. ResNet20) using 4 GPUs - to achieve high GPU utilization (with room for improvement).

File: examples/classifier_compression/parallel-finetune.py [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+            f.flush()
+
+
+class FTStatsLogger(_CSVLogger):
+    def __init__(self, fname):
+        headers = ['dir', 'name', 'macs', 'search_top1', 'top1', 'top5', 'loss']
+        super().__init__(fname, headers)
+
+
+class FinetuningTask(object):
```
[the remark this answers is not in the document]
Comment (nzmora) [reply examples/classifier_compression/parallel-finetune.py]: I would use 'parallel-fine-tune' or 'parallel-fine-tuning' instead of camel-case. 
But changing the name requires changing the examples in the notebooks, so for now I'm leaving it as is.

File: examples/classifier_compression/compress_classifier.py (hunk 1 above)
Comment (nzmora) [reply examples/classifier_compression/compress_classifier.py]: done

File: examples/classifier_compression/compress_classifier.py (hunk 2 above)
Comment (nzmora) [reply examples/classifier_compression/compress_classifier.py]: done
